packages/core/src/storages/request_provider.ts

```diff
@@ -0,0 +1,755 @@
+import { ListDictionary, LruCache } from '@apify/datastructures';
+import type { Log } from '@apify/log';
+import { cryptoRandomObjectId } from '@apify/utilities';
+import type {
+    BatchAddRequestsResult,
+    Dictionary,
+    ProcessedRequest,
+    QueueOperationInfo,
+    RequestQueueClient,
+    RequestQueueInfo,
+    StorageClient,
+} from '@crawlee/types';
+import { chunk, downloadListOfUrls, sleep } from '@crawlee/utils';
+import ow from 'ow';
+
+import type { IStorage, StorageManagerOptions } from './storage_manager';
+import { StorageManager } from './storage_manager';
+import { QUERY_HEAD_MIN_LENGTH, STORAGE_CONSISTENCY_DELAY_MILLIS, getRequestId, purgeDefaultStorages } from './utils';
+import { Configuration } from '../configuration';
+import { EventType } from '../events';
+import { log } from '../log';
+import type { ProxyConfiguration } from '../proxy_configuration';
+import { Request } from '../request';
+import type { RequestOptions, InternalSource, Source } from '../request';
+import type { Constructor } from '../typedefs';
+
+export abstract class RequestProvider implements IStorage {
+    id: string;
+    name?: string;
+    timeoutSecs = 30;
+    clientKey = cryptoRandomObjectId();
+    client: RequestQueueClient;
+    protected proxyConfiguration?: ProxyConfiguration;
+
+    log: Log;
+    internalTimeoutMillis = 5 * 60_000; // defaults to 5 minutes, will be overridden by BasicCrawler
+
+    // We can trust these numbers only in a case that queue is used by a single client.
+    // This information is returned by getHead() under the hadMultipleClients property.
+    assumedTotalCount = 0;
+    assumedHandledCount = 0;
+
+    protected queueHeadIds = new ListDictionary<string>();
+    protected requestCache: LruCache<RequestLruItem>;
+    /** @internal */
+    inProgress = new Set<string>();
+    protected recentlyHandledRequestsCache: LruCache<boolean>;
+
+    // TODO: RQv1 logic for stuck queues, this might not be needed anymore
+    protected lastActivity = new Date();
+    protected queuePausedForMigration = false;
+
+    constructor(options: InternalRequestProviderOptions, readonly config = Configuration.getGlobalConfig()) {
+        this.id = options.id;
+        this.name = options.name;
+        this.client = options.client.requestQueue(this.id, {
+            clientKey: this.clientKey,
+            timeoutSecs: this.timeoutSecs,
+        });
+
+        this.proxyConfiguration = options.proxyConfiguration;
+
+        this.requestCache = new LruCache({ maxLength: options.requestCacheMaxSize });
+        this.recentlyHandledRequestsCache = new LruCache({ maxLength: options.recentlyHandledRequestsMaxSize });
+        this.log = log.child({ prefix: options.logPrefix });
+
+        const eventManager = config.getEventManager();
```

Comment:
Register events for migration is just for rqV2; I would not mix it here.

Reply:
what about this, i see its resolved by nothing changed, not sure if there was some discussion outside of GH or its just forgotten?